CONTRIBUTING.md

```diff
@@ -0,0 +1,76 @@
+# Contributing
+
+## Overview
+
+This document explains the processes and practices recommended for contributing enhancements to
+this operator.
+
+- Generally, before developing enhancements to this charm, you should consider [opening an issue
+  ](https://github.com/canonical/training-operator/issues) explaining your use case.
+- If you would like to chat with us about your use-cases or proposed implementation, you can reach
+  us at [Canonical Mattermost public channel](https://chat.charmhub.io/charmhub/channels/charm-dev)
+  or [Discourse](https://discourse.charmhub.io/).
+- Familiarising yourself with the [Charmed Operator Framework](https://juju.is/docs/sdk) library
+  will help you a lot when working on new features or bug fixes.
+- All enhancements require review before being merged. Code review typically examines
+  - code quality
+  - test coverage
+  - user experience for Juju administrators of this charm.
+- Please help us out in ensuring easy to review branches by rebasing your pull request branch onto
+  the `main` branch. This also avoids merge commits and creates a linear Git commit history.
+
+## Developing
+
+You can use the environments created by `tox` for development:
+
+```shell
+tox --notest -e unit
+source .tox/unit/bin/activate
+```
+
+### Testing
+
+```shell
+tox -e lint          # code style
+tox -e unit          # unit tests
+tox -e integration   # integration tests
+tox                  # runs 'lint' and 'unit' environments
+```
+
+## Build charm
+
+Build the charm in this git repository using:
+
+```shell
+charmcraft pack
+```
+
+### Deploy
+
+```bash
+# Create a model
+juju add-model dev
+# Enable DEBUG logging
+juju model-config logging-config="<root>=INFO;unit=DEBUG"
+# Deploy the charm
+juju deploy ./training-operator_ubuntu-20.04-amd64.charm \
+    --resource training-operator-image=kubeflow/training-operator
+```
+
+## Charmed Training Operator source and resources upgrade
+
+To upgrade the source and resources Charmed Training Operator, you must:
+
+1) Bump the `training-operator-image` in `metadata.yaml`
+2) Update all manifests in `src/` to match those in the version you want to upgrade to
+3) Ensure integration and unit tests are passing; fix/adapt them otherwise
+
+The following table collects a series of links and information that can help in the upgrade process.
+ 
+| Image repository 	| Image version 	| Manifest location 	| Next version 	|
```

Comment:
I wasn't sure what the `image version` column is supposed to convey.  Is this the format of how they tag their images?

Reply:
Actually I'm not sure what storing `image repository` here adds.  Is that something that someone trying to update the charm will want?  I'd expect they'd use whatever the training operator release page said

Reply:
Do we want to cite here the manifests from the workload's repo, or from `kubeflow/manifests`?  I'm torn.  We usually follow `kubeflow/manifests`, although it feels funny to put that here instead of the workload's repo.  But for something like kfp, going to their repo and looking for releases is probably misleading for our purposes

Reply:
> I wasn't sure what the `image version` column is supposed to convey. Is this the format of how they tag their images?

It is a reference to make it easier to identify which image belongs to which release.

Reply:
> Actually I'm not sure what storing `image repository` here adds. Is that something that someone trying to update the charm will want? I'd expect they'd use whatever the training operator release page said

How do you know where to look? I always struggle when trying to find the registry where each image is hosted. I don't see that info in the [release page](https://github.com/kubeflow/training-operator/releases/tag/v1.5.0-rc.0).

Reply:
I should have put these three threads as comments on separate lines of the table :p  I'm splitting them up...

Re image version column:
I'm not sure what someone coming to this repo gains by having this info.  I'm picturing them going to the upstream manifest, looking for the images, and seeing there's new tags on the end.  I don't know what they'd do with the format of those tags beforehand?  

I have a feeling I might be misunderstanding something.

Reply:
continuing the conversation about `image repository` [here](https://github.com/canonical/training-operator/pull/33/files#r922450851)

Reply:
continuing the conversation about `kubeflow/manifests` vs workload repo [here](https://github.com/canonical/training-operator/pull/33/files#r922458659)

Reply:
> I'm not sure what someone coming to this repo gains by having this info. I'm picturing them going to the upstream manifest, looking for the images, and seeing there's new tags on the end. I don't know what they'd do with the format of those tags beforehand?

Where would they look for? At least for me, the experience has always been time consuming for not knowing this because I do not know if there's a file with this info. If you have it, we can link it here. 
Also, it is convenient to know the format of the tag, it makes tags predictable for automation, makes it easy to catch errors if for some reason the image tag in a PR/change doesn't match what we expect, it is easier to go into the image registry and easily find images and relate them to releases, makes it easier for someone debugging differences between images.

Reply:
I think the tag format is very interesting, I didn't know training-operator formated it this way. I would rename this column from `image version` to `image tag format`.

Reply:
I think this is hashed out more in the other threads now, but I look at a manifest for everything.  The image repo, even if I know how versions are formatted, doesn't tell me which version I should use.  That info is in the manifests

Agreed if we're building automation around the image, knowing the format of the tag is critical.  But we aren't doing that right now, so I don't think that's a good justification on its own to bother now.  If we automated later, we'd do the work then

Reply:
Given that we are sticking with linking both manifest repositories, I don't think having this column hurts. Let's keep it in and resolve this.